### PR TITLE
[feat] init kornia.nn

### PR DESCRIPTION
**Pull request by shijianjian**
Added the auto nn builder with a module registry.

We can also add some custom field as a record like:
```python
@nn.register_module(differentiable=True, jitable=False)
class RandomFlip(...)
```

/cc @lferraz

### CODE REVIEW

**Review by edgarriba — COMMENTED** [review]
@shijianjian can we easily turn functionals to modules with this ? Or what's the actual use case for that?

**Comment by shijianjian** [comment]
This is a nn module wrapper to include all the modules that can be used as neural network layers. This is a more or less AOP idea to collect and improve those layers. Essentially:

1. we can wrap in all the layer-wise routine functions, like differentiability-check, onnx-check, jitable-check. 
2. do the signature inspect stuff as in the future limbus usage.
3. use jinja2 to alter the modules at the runtime.
4. users can quickly go through what they could use to compose networks.

**Comment by edgarriba** [comment]
> 1. we can wrap in all the layer-wise routine functions, like differentiability-check, onnx-check, jitable-check.
> 
>     2. do the signature inspect stuff as in the future limbus usage.
> 
>     3. use jinja2 to alter the modules at the runtime.
> 
>     4. users can quickly go through what they could use to compose networks.

Perfect -- let's add this notes to the docs of the module. I'd also suggest the following, to create a base class `Algorithm` (seamless to opencv main base class) so that all the nn.Module version of the functionals can inherit so we can easily automate specific functionalites e.g `to_onxx`, `to_jit`, `serialise`, `deserialise` and a long etc.

```python
class Algorithm(nn.Module):
    accelerator = Accelerator()  # possibly in the future from torch.x

    def preprocess(self, image: Tensor) -> Tensor:
        raise NotImplementedError

    def postprocess(self, image: Tensor) -> Dict[str, Tensor]:
        raise NotImplementedError

    # TODO: figure out what happens when more than one input image comes,
    # e.g in optical flow or stereo, image matching, etc.
    def forward(self, image: Tensor) -> Dict[str, Tensor]:
        raise NotImplementedError

    def finetune(self, config: TrainerConfiguration, augmentations: nn.Module = None):
        # define implementation -- and not all the algorithms will need this
        ...

    def save(self, path_file: str):
        torch.save(self.model, path_file)

    @classmethod
    def from_filepath(cls, path_file: str):
        return cls(torch.load(path_file))

    def to_jit(self):
        raise NotImplementedError("not implemented yet")

    def to_onnx(self, filepath: str, input_data: torch.Tensor):
        self(example_input)
        torch.onnx.export(
            self,  # model being run
            example_input,  # model input (or a tuple for multiple inputs)
            filepath,  # where to save the model (can be a file or file-like object)
            export_params=True,  # store the trained parameter weights inside the model file
            opset_version=11,  # the ONNX version to export the model to
            do_constant_folding=True,  # whether to execute constant folding for optimization
            input_names=["input"],  # the model's input names
            output_names=["output"],  # the model's output names
            dynamic_axes={
                "input": {0: "batch_size"},  # variable length axes
                "output": {0: "batch_size"},
            },
        )
```
Then
```python
class GaussianBlur2d(Algorithm):
    # implement whatever is needed
     ....
```

**Comment by shijianjian** [comment]
The ```preprocessing``` and ```postprocessing``` should happen in containers like an ```ImageSequential```. For the ```Algorithm``` part, I would like to see it happen. With a clean usage like:

```python
method = Algorithm(
   K.AugmentationSequential(...),
   K.LofTR(...)
)
method.finetune()
method.to_onnx(...)
```

What defined in this PR is much simpler. Things like tag on a class properties (e.g. method.jitterable = False), add routine (common) docs, keep track on all the modules etc. More lightweight without much functionalities.

File: docs/source/nn.rst [hunk 1]
```diff
@@ -32,35 +32,35 @@ Augmentation Layers
...
-    kornia.nn.RandomResizedCrop
-    kornia.nn.RandomRotation
-    kornia.nn.RandomSharpness
-    kornia.nn.RandomSolarize
-    kornia.nn.RandomThinPlateSpline
-    kornia.nn.RandomVerticalFlip
-    kornia.nn.PadTo
-    kornia.nn.Normalize
-    kornia.nn.Denormalize
+    nn.RandomHorizontalFlip
```
Comment (edgarriba) [inline docs/source/nn.rst:35]: Are we going to have duplicate content for augmentations?

Comment (shijianjian) [reply docs/source/nn.rst:35]: Only titles. This will generate a list than full docs.

Comment (edgarriba) [reply docs/source/nn.rst:35]: Should we remove then any module-like from the existing docs ?

**Review by edgarriba — COMMENTED** [review]
Is there any minimal test that we can add here ?

File: docs/source/nn.rst (hunk 1 above)
Comment (shijianjian) [reply docs/source/nn.rst:35]: Not sure on that. I would take the nn modules as a shortcut.

**Comment by shijianjian** [comment]
> Is there any minimal test that we can add here ?

Probably. I will investigate what are worth testing.

**Pull request closed without merging**